op-service/eth/transactions.go

```diff
@@ -69,25 +71,47 @@ func CheckRecentTxs(
 		return 0, false, fmt.Errorf("failed to retrieve current nonce: %w", err)
 	}
 
-	oldestBlock := new(big.Int)
-	oldestBlock.Sub(currentBlock, big.NewInt(int64(depth)))
+	oldestBlock := new(big.Int).Sub(currentBlock, big.NewInt(int64(depth)))
 	previousNonce, err := l1.NonceAt(ctx, addr, oldestBlock)
 	if err != nil {
 		return 0, false, fmt.Errorf("failed to retrieve previous nonce: %w", err)
 	}
 
-	if currentNonce == previousNonce {
+	if currentNonce <= previousNonce {
 		return oldestBlock.Uint64(), false, nil
 	}
 
-	// Decrease block num until we find the block before the most recent batcher tx was sent
-	targetNonce := currentNonce - 1
-	for currentNonce > targetNonce && currentBlock.Cmp(oldestBlock) != -1 {
-		currentBlock.Sub(currentBlock, big.NewInt(1))
-		currentNonce, err = l1.NonceAt(ctx, addr, currentBlock)
+	// Use binary search to find the block where the nonce changed
+	low := oldestBlock.Uint64()
+	high := currentBlock.Uint64()
+
+	for low < high {
+		mid := (low + high) / 2
+		midNonce, err := l1.NonceAt(ctx, addr, new(big.Int).SetUint64(mid))
 		if err != nil {
-			return 0, false, fmt.Errorf("failed to retrieve nonce: %w", err)
+			return 0, false, fmt.Errorf("failed to retrieve nonce at block %d: %w", mid, err)
+		}
+
+		if midNonce > currentNonce {
+			// Catch a reorg that causes inconsistent nonce
+			return oldestBlock.Uint64(), false, nil
```

Comment:
I'm not sure this is how we want to handle a reorg case. If the mid nonce is _higher_ than the pre-reorg current nonce, it means that there suddenly may be a more recent transaction between oldestBlock and the currentBlock. I think at this point we cannot narrow it down to a smaller range. So maybe what we want to do instead here is to restart a new search for the same range? Easiest would probably be again to just recursively call `CheckRecentTxs` because we'd also need to update the `currentNonce`.
```suggestion
		if midNonce > currentNonce {
			// Restart search on reorg
			return CheckRecentTxs(ctx, l1, depth, addr)
```

Reply:
https://github.com/ethereum-optimism/optimism/pull/11232/commits/9f6009764310d8a35f91b2cbe9fa5281f8616eb2